uniffi_bindgen/src/interface/mod.rs

```diff
@@ -241,6 +241,42 @@ impl<'ci> ComponentInterface {
         }
     }
 
+    /// Identify if we are using an unsigned type in kotlin
+    ///
+    /// In Kotlin, unsigned types are expiremental and will cause a warning if used without annotations
+    /// This should be removed when kotlin version
+    ///  > v1.5 https://kotlinlang.org/docs/whatsnew15.html#stable-unsigned-integer-types
+    pub fn contains_unsigned_type(&self, type_: &Type) -> bool {
+        match type_ {
+            Type::UInt8 | Type::UInt16 | Type::UInt32 | Type::UInt64 => true,
+            Type::Optional(t) | Type::Sequence(t) | Type::Map(t) => self.contains_unsigned_type(t),
+            Type::Record(name) => self
+                .get_record_definition(name)
+                .map(|rec| {
+                    rec.fields()
+                        .iter()
+                        .any(|f| self.contains_unsigned_type(&f.type_))
+                })
```

Comment:
I find myself wondering if this "iterate over each field" logic should live in `Record::contains_unsigned_type` struct and we just call `rec.contains_unsigned_type(self)` here, rather than having this `ComponentInterface::contains_unsigned_type` method know too many details about the internals of the `Record` struct. But I note that the existing `ComponentInterface::type_contains_object_references` uses the same pattern as you've got here, so my past self clearly didn't think it was too big of a deal 😅.

If you do decide to add similar helper methods to the `Function`/`Constructor`/`Method` structs in addition to `Enum` and `Record`, then it's probably worth pushing the implementation details for `Enum` and `Record` out into those structs for consistency, rather than having them inline here.

Reply:
hmm that's an interesting pattern -- I definitely modeled it after the `type_contains_object_references` in which the structs are actually simply calling the `ComponentInterface` under the hood (rather than CI calling the respective structs itself). However I think that may be the way to approach it as you're right it could abstract away the implementation of these different types, not to mention adding new types is more straightforward. I think I'll give this a shot.